---
title: "scCancer"
author: "wguo"
date: "2019/6/13"
output: html_document
---

<style type="text/css">
    body{
        font-size: 15px;
        line-height: 22px;
    }
    h1.title {
        font-size: 38px;
    }
    h1 {
        font-size: 28px;
        margin-top: 23px;
    }
    h2 {
        font-size: 24px;
        margin-top: 25px;
    }
    h3 {
      font-size: 20px;
        margin-top: 25px;
    }
    code.r{
        font-size: 13px;
    }
    pre {
        font-size: 14px;
    }
    p {
        margin-top:10px;
        margin-bottom:10px;
    }
    table { 
        width: 60%;
        border-collapse: collapse;
        font-family: Futura, Arial, sans-serif;
    }
    th,td {
        padding: 5px;
    }
    th,td {
        border-bottom: 1px solid #ddd;
        border-top: 1px solid #ddd;
        padding-right: 20px
    }
</style>



```{r setting, include=FALSE}
options(knitr.table.format = "html") 
options(scipen=10)
knitr::opts_chunk$set(echo = TRUE, fig.path = file.path(results$savePath, 'report-figures//'))

title <- "scCancer"
if(!is.null(results$sampleName)){
  title <- paste0(results$sampleName, "  -  ", title)
}

if(!is.null(results$authorName)){
  userName <- results$authorName
}else{
  userName <- Sys.getenv("USERNAME")
}
reportMark <- Sys.time()
if(userName != ""){
  reportMark <- paste0(userName, " , ", reportMark)
}

h.i <- 1
h.ii <- 1
```


# `r title`
--------------------------------
<p align="right">`r reportMark`</p>


## `r h.i` Cell statistics

* The input of `scCancer` pipeline is the matrix generated by [` `r results$cr.version`  `](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger).

```{r echo=F, results='asis'}
if(file.exists(file.path(results$dataPath, "web_summary.html"))){
  file.copy(file.path(results$dataPath, "web_summary.html"), file.path(results$savePath, "report-cellRanger.html"), overwrite = T)
  cat("* Here is the [summary report](./report-cellRanger.html) from `Cell Ranger`.", sep = "")
}
```


### `r h.i`.`r h.ii` Cell calling
 
```{r CR-Calling-T, child=system.file("rmd", "cellCalling.Rmd", package = "scCancer"), eval = results$raw.data}
```

```{r CR-Calling-F, echo=FALSE, results='asis', eval=!results$raw.data}
cat("* Raw data (containing all barcodes) cannot be found, and only filtered data are supplied. So cell-calling doesn't be performed and the analyses for background distribution is omitted.\n")
cat("* For the filtered data, ", results$nList[2], " cells are identified (min.nUMI = `", results$min.nUMI, "`).\n", sep="")
```

```{r echo=F}
h.ii <- h.ii + 1
# print(results$raw.data)
```




### `r h.i`.`r h.ii` The number of UMIs and detected genes in cells

After the cell calling by ` `r results$cr.version` `, we further perform quality control to 
filter droplets with low quality cells according to `nUMI` (total number of UMIs) and `nGene` (total number of detected genes).

For `nUMI` :
* Suggested threshold to filter cells with extremely large `nUMI` : <span style="color:red">` `r results$cell.threshold$nUMI` `</span>.
    + Using this threshold, ` `r sum(results$cell.manifest$nUMI >= results$cell.threshold$nUMI)` ` cells will be filtered.

For `nGene` :
* Suggested threshold to filter cells with extremely large `nGene` : <span style="color:red">` `r results$cell.threshold$nGene` `</span>.
    + Using this threshold, ` `r sum(results$cell.manifest$nGene >= results$cell.threshold$nGene)` ` cells will be filtered.
* Suggested threshold to filter cells with extremely small `nGene` : <span style="color:red">`200`</span>.
    + Using this threshold, ` `r sum(results$cell.manifest$nGene < 200)` ` cells will be filtered.

**Comment**: The suggested thresholds (except the lower bound of `nGene`, which is set by convention) are determined based on the their distributions. Using them, the outliers identified will be filtered. The same below.


```{r filter, echo=F, message=F, warning=F, dpi=300, fig.height=2.5, fig.width=8}
plot_grid(results$p.nUMI, results$p.nGene, ncol = 2)
```
<p align='right' style='margin-top:3px'>(Hi-res image: <a href='./figures/nUMI-distr.png'>left</a>, <a href='./figures/nGene-distr.png'>right</a>)</p>


```{r echo=F}
h.i <- h.i + 1
h.ii <- 1
```







## `r h.i` Gene statistics
The number of genes expressed in at least one cell : ` `r sum(results$gene.manifest$nCell > 0)` `.


### `r h.i`.`r h.ii` Mitochondrial genes
Summary of mitochondrial genes percentage (`mito.percent`) in cells:
```{r mito.summary, echo=F}
format(summary(results$cell.manifest$mito.percent), digits = 3)
```
* Suggested threshold to filter cells with high mitochondrial genes percentage : <span style="color:red">` `r round(results$cell.threshold$mito.percent, 3)` `</span>.
    + Using this threshold, ` `r sum(results$cell.manifest$mito.percent >= results$cell.threshold$mito.percent)` ` cells will be filtered.

```{r mito, echo=FALSE, message=F, dpi=300, fig.height=4, fig.width=4, fig.align="center", out.width='40%'}
results$p.mito
```
<p align="right">(Hi-res image: <a href="./figures/mito-distr.png">view</a>)</p>


```{r echo=F}
h.ii <- h.ii + 1
```



### `r h.i`.`r h.ii` Ribosome genes
Summary of ribosome genes percentage (`ribo.percent`) in cells:
```{r ribo.summary, echo=F}
format(summary(results$cell.manifest$ribo.percent), digits = 3)
```
* Suggested threshold to filter cells with high ribosome genes percentage : <span style="color:red">` `r round(results$cell.threshold$ribo.percent, 3)` `</span>.
    + Using this threshold, ` `r sum(results$cell.manifest$ribo.percent >= results$cell.threshold$ribo.percent)` ` cells will be filtered.

```{r ribo, echo=FALSE, message=F, dpi=300, fig.height=4, fig.width=4, fig.align="center", out.width='40%'}
results$p.ribo
```
<p align="right">(Hi-res image: <a href="./figures/ribo-distr.png">view</a>)</p>

```{r echo=F}
h.ii <- h.ii + 1
```



### `r h.i`.`r h.ii` Dissociation associated genes
Summary of dissociation associated genes percentage (`diss.percent`) in cells:
```{r diss.summary, echo=F}
format(summary(results$cell.manifest$diss.percent), digits = 3)
```
* Suggested threshold to filter cells with high dissociation genes percentage : <span style="color:red">` `r round(results$cell.threshold$diss.percent, 3)` `</span>.
    + Using this threshold, ` `r sum(results$cell.manifest$diss.percent >= results$cell.threshold$diss.percent)` ` cells will be filtered.

```{r diss, echo=FALSE, message=F, dpi=300, fig.height=4, fig.width=4, fig.align="center", out.width='40%'}
results$p.diss
```
<p align="right">(Hi-res image: <a href="./figures/diss-distr.png">view</a>)</p>

```{r echo=F}
h.ii <- h.ii + 1
```




### `r h.i`.`r h.ii` Ambient RNAs 

### `r h.i`.`r h.ii`.1 Highly-expressed genes
In order to analyze the gene expression profiles in detail and identify  
highly-expressed genes in background mRNAs from lysed cells, 
we calculate some metrics as shown below.
```{r echo=F, results='asis'}
if("bg.percent" %in% colnames(results$gene.manifest)){
  cat("* `bg.percent` : the expression proportion for each gene in background distribution (all droplets with `nUMI <= 10`).\n")
}
```
* `prop.median`	: the median of expression proportions for a gene in each cell.
* `detect.rate`	: the detected (`#UMI > 0`) rate for a gene in all cells.

Here is a plot showing the distributions of gene proportion in cells for the first 100 genes (ordered by their proportion in background `bg.percent`). And the points (genes) are colored according to whether they belongs to mitochondrial, ribosome, or dissociation associated genes.
```{r echo=F, results='asis'}
if("bg.percent" %in% colnames(results$gene.manifest)){
  cat("The red star signs mark the genesâ€™ proportion in background.\n")
}
```

```{r genePropPlot, echo=F, message=F, warning=F, dpi=300, fig.width=8, fig.height=8, fig.align="center"}
grid::grid.draw(results$p.geneProp)
```
<p align="right">(Hi-res image: <a href="./figures/geneProp.png">view</a>)</p>


```{r echo=F, results='asis'}
if("bg.percent" %in% colnames(results$gene.manifest)){
  cat("The plot below shows the relationship between `bg.percent` and `prop.median`, `bg.percent` and `detect.rate`.\n")
}
```

```{r gene.plot, echo=F, message=F, warning=F, dpi=300, fig.height=4, fig.width=8}
if("bg.percent" %in% colnames(results$gene.manifest)){
  plot_grid(results$p.bg.cell, results$p.bg.detect, ncol = 2)
}
```

```{r echo=F, results='asis'}
if("bg.percent" %in% colnames(results$gene.manifest)){
  cat("<p align=\"right\">(Hi-res image: <a href=\"./figures/bg-cell-scatter.png\">left</a>, <a href=\"./figures/bg-detect-scatter.png\">right</a>)</p>\n")
  # cat("(Hi-res image\n")
}
```

```{r echo=F}
h.i <- h.i + 1
h.ii <- 1
```



```{r soupx, child=system.file("rmd", "SoupX.Rmd", package = "scCancer"), eval = results$bool.runSoupx}
```




## `r h.i` Output

### `r h.i`.`r h.ii` Thresholds to filter droplets
According to the results of statistics and visualization, we propose following thresholds to filter cells:

<center>
```{r thresholds, echo=F, warning=F}
# results$filter.thres %>% knitr::kable("html")
kable(results$filter.thres)
```
</center>


* **Hint**: In general, `Cell Ranger` can filter the droplets with low nUMI. So here we set `Low.threshold` for nUMI as `0`.
The users need to use the identification results of `Cell Ranger` or set a suitable threshold first to filter the possible empty droplets with less UMIs.


Using these thresholds, the number of cells vary as follows:

` `r paste0("Raw : ", results$nList[1])` ` ->
` `r paste0("cellranger3 : ", results$nList[2])` ` ->
` `r paste0("nUMI<", results$cell.threshold$nUMI, 3, " : ", results$nList[3])` ` ->
` `r paste0("nGene>=200 : ", results$nList[4])` ` ->
` `r paste0("nGene<", results$cell.threshold$nGene, " : ", results$nList[5])` ` ->
` `r paste0("mito.percent<", round(results$cell.threshold$mito.percent, 3), " : ", results$nList[6])` ` ->
` `r paste0("ribo.percent<", round(results$cell.threshold$ribo.percent, 3), " : ", results$nList[7])` ` ->
` `r paste0("diss.percent<", round(results$cell.threshold$diss.percent, 3), " : ", results$nList[8])` `

```{r echo=F}
h.ii <- 1
```


### `r h.i`.`r h.ii` Output files
Running this script generates following files:

```{r echo=F}
r.i <- 8
```

1. **Html report** :
[report-scStat.html](./report-scStat.html).
2. **Markdown report** :
[report-scStat.md](./report-scStat.md).
3. **Figure files** :
[figures/](./figures/).
4. **Figures used in the report**:
[report-figures/](./report-figures/).
5. **Text file with cell manifest** :
[cellManifest-all.txt](./cellManifest-all.txt).
6. **Text file with suggested thresholds as above** :
[cell.QC.thres.txt](./cell.QC.thres.txt).
7. **Text file with gene manifest** :
[geneManifest.txt](./geneManifest.txt).
```{r echo=F, results='asis',  eval=results$bool.runSoupx}
cat(r.i, ". **Text file with the results of SoupX** :[ambientRNA-SoupX.txt](./ambientRNA-SoupX.txt).", 
    sep = "")
r.i <- r.i + 1
```
```{r echo=F, results='asis'}
if(file.exists(file.path(results$dataPath, "web_summary.html"))){
  cat("9. **Cell ranger html report** (Copy from the source data folder):\n")
  cat("[report-cellRanger.html](./report-cellRanger.html).\n", sep = "")
}
```



<br>

--------------------------------------
&copy; [G-Lab](http://lifeome.net/glab/jgu/),   [Tsinghua University](http://www.tsinghua.edu.cn)
